File: scripts/test_batch_100.js
/**
 * QSM Batch Test Script - 100 Documents
 * 
 * Automatically selects and processes 100 diverse documents from test folder
 * Measures: processing time, memory usage, success rate, error patterns
 */

const fs = require('fs-extra');
const path = require('path');
const { processDocument } = require('../dist/utils/documentProcessor');
const { performance } = require('perf_hooks');

// Configuration
const TEST_FOLDER = 'D:\\Work\\Coding\\archi-query-master\\Documents';
const RESULT_FILE = path.join(__dirname, '..', 'test_results_100.json');
const LOG_FILE = path.join(__dirname, '..', 'test_log_100.txt');
const TARGET_COUNT = 100;

// Test configuration
const config = {
  chunkSize: 320,
  chunkOverlap: 50
};

// Results tracking
const results = {
  startTime: new Date().toISOString(),
  config,
  totalFiles: 0,
  processedFiles: 0,
  failedFiles: 0,
  totalProcessingTime: 0,
  avgProcessingTime: 0,
  files: [],
  errors: [],
  summary: {}
};

// Logging utility
function log(message) {
  const timestamp = new Date().toISOString();
  const logMessage = `[${timestamp}] ${message}\n`;
  console.log(message);
  fs.appendFileSync(LOG_FILE, logMessage);
}

// Select diverse files
async function selectTestFiles() {
  log(`📂 Scanning ${TEST_FOLDER}...`);
  
  const allFiles = [];
  
  // Recursively get all files
  function scanDir(dir) {
    const items = fs.readdirSync(dir);
    for (const item of items) {
      const fullPath = path.join(dir, item);
      const stat = fs.statSync(fullPath);
      
      if (stat.isDirectory()) {
        scanDir(fullPath);
      } else {
        const ext = path.extname(item).toLowerCase();
        if (['.pdf', '.docx', '.doc', '.txt', '.xlsx', '.pptx'].includes(ext)) {
          allFiles.push({
            path: fullPath,
            name: item,
            size: stat.size,
            ext
          });
        }
      }
    }
  }
  
  scanDir(TEST_FOLDER);
  
  log(`✅ Found ${allFiles.length} documents`);
  
  // Select diverse sample (stratified by size and type)
  const selected = [];
  
  // Group by type
  const byType = {
    '.pdf': allFiles.filter(f => f.ext === '.pdf'),
    '.docx': allFiles.filter(f => f.ext === '.docx'),
    '.doc': allFiles.filter(f => f.ext === '.doc'),
    '.txt': allFiles.filter(f => f.ext === '.txt'),
    '.xlsx': allFiles.filter(f => f.ext === '.xlsx'),
    '.pptx': allFiles.filter(f => f.ext === '.pptx')
  };
  
  // Distribute proportionally
  const totalFiles = allFiles.length;
  for (const [ext, files] of Object.entries(byType)) {
    if (files.length === 0) continue;
    
    const proportion = files.length / totalFiles;
    const targetCount = Math.max(1, Math.floor(TARGET_COUNT * proportion));
    
    // Sample evenly across size ranges
    files.sort((a, b) => a.size - b.size);
    const step = Math.max(1, Math.floor(files.length / targetCount));
    
    for (let i = 0; i < files.length && selected.length < TARGET_COUNT; i += step) {
      selected.push(files[i]);
    }
  }
  
  // Fill remaining with random selection if needed
  while (selected.length < TARGET_COUNT && selected.length < allFiles.length) {
    const remaining = allFiles.filter(f => !selected.includes(f));
    if (remaining.length === 0) break;
    
    const randomIndex = Math.floor(Math.random() * remaining.length);
    selected.push(remaining[randomIndex]);
  }
  
  log(`✅ Selected ${selected.length} files for testing`);
  log(`📊 Distribution: ${JSON.stringify(
    selected.reduce((acc, f) => {
      acc[f.ext] = (acc[f.ext] || 0) + 1;
      return acc;
    }, {})
  )}`);
  
  return selected;
}

// Process single file
async function processFile(file, index, total) {
  const startTime = performance.now();
  const startMemory = process.memoryUsage().heapUsed;
  
  log(`\n[${index + 1}/${total}] Processing: ${file.name} (${(file.size / 1024 / 1024).toFixed(2)} MB)`);
  
  try {
    // Create File object (simulate browser File API)
    const buffer = await fs.readFile(file.path);
    const fileObj = {
      name: file.name,
      size: file.size,
      type: file.ext === '.pdf' ? 'application/pdf' : 'application/octet-stream',
      arrayBuffer: async () => buffer
    };
    
    // Process document
    const result = await processDocument(
      fileObj,
      file.name,
      (progress) => {
        if (progress % 25 === 0) {
          log(`  ⏳ Progress: ${progress}%`);
        }
      },
      null, // No embedding pipeline for test
      config.chunkSize,
      config.chunkOverlap
    );
    
    const endTime = performance.now();
    const endMemory = process.memoryUsage().heapUsed;
    const processingTime = (endTime - startTime) / 1000; // seconds
    const memoryUsed = (endMemory - startMemory) / 1024 / 1024; // MB
    
    log(`  ✅ SUCCESS - ${processingTime.toFixed(2)}s, ${result.chunks} chunks, ${result.metadata.wordCount} words`);
    
    results.processedFiles++;
    results.totalProcessingTime += processingTime;
    results.files.push({
      name: file.name,
      path: file.path,
      size: file.size,
      ext: file.ext,
      status: 'success',
      processingTime,
      memoryUsed,
      chunks: result.chunks,
      wordCount: result.metadata.wordCount,
      pageCount: result.metadata.pageCount,
      tableCount: result.metadata.tableCount,
      confidence: result.metadata.confidence
    });
    
    return true;
  } catch (error) {
    const endTime = performance.now();
    const processingTime = (endTime - startTime) / 1000;
    
    log(`  ❌ FAILED - ${error.message}`);
    
    results.failedFiles++;
    results.files.push({
      name: file.name,
      path: file.path,
      size: file.size,
      ext: file.ext,
      status: 'failed',
      processingTime,
      error: error.message
    });
    
    results.errors.push({
      file: file.name,
      error: error.message,
      stack: error.stack
    });
    
    return false;
  }
}

// Main test runner
async function runBatchTest() {
  log('🚀 Starting QSM Batch Test - 100 Documents\n');
  log(`📋 Configuration: chunk_size=${config.chunkSize}, chunk_overlap=${config.chunkOverlap}\n`);
  
  // Select files
  const files = await selectTestFiles();
  results.totalFiles = files.length;
  
  log(`\n${'='.repeat(60)}`);
  log('🔄 STARTING BATCH PROCESSING');
  log('='.repeat(60) + '\n');
  
  // Process files sequentially (avoid memory issues)
  for (let i = 0; i < files.length; i++) {
    await processFile(files[i], i, files.length);
    
    // Memory cleanup
    if (global.gc && i % 10 === 0) {
      global.gc();
    }
  }
  
  // Calculate summary
  results.endTime = new Date().toISOString();
  results.avgProcessingTime = results.totalProcessingTime / results.processedFiles;
  
  results.summary = {
    successRate: `${((results.processedFiles / results.totalFiles) * 100).toFixed(2)}%`,
    avgProcessingTime: `${results.avgProcessingTime.toFixed(2)}s`,
    totalTime: `${(results.totalProcessingTime / 60).toFixed(2)} minutes`,
    fastestFile: results.files.filter(f => f.status === 'success').sort((a, b) => a.processingTime - b.processingTime)[0],
    slowestFile: results.files.filter(f => f.status === 'success').sort((a, b) => b.processingTime - a.processingTime)[0],
    byType: {}
  };
  
  // Statistics by file type
  for (const ext of ['.pdf', '.docx', '.doc', '.txt', '.xlsx', '.pptx']) {
    const typeFiles = results.files.filter(f => f.ext === ext);
    if (typeFiles.length > 0) {
      const successCount = typeFiles.filter(f => f.status === 'success').length;
      results.summary.byType[ext] = {
        total: typeFiles.length,
        success: successCount,
        failed: typeFiles.length - successCount,
        avgTime: typeFiles
          .filter(f => f.status === 'success')
          .reduce((sum, f) => sum + f.processingTime, 0) / successCount || 0
      };
    }
  }
  
  // Save results
  await fs.writeJson(RESULT_FILE, results, { spaces: 2 });
  
  // Print summary
  log(`\n${'='.repeat(60)}`);
  log('📊 TEST SUMMARY');
  log('='.repeat(60));
  log(`Total Files:      ${results.totalFiles}`);
  log(`Processed:        ${results.processedFiles} ✅`);
  log(`Failed:           ${results.failedFiles} ❌`);
  log(`Success Rate:     ${results.summary.successRate}`);
  log(`Total Time:       ${results.summary.totalTime}`);
  log(`Avg Time/File:    ${results.summary.avgProcessingTime}`);
  log(`Fastest:          ${results.summary.fastestFile?.name} (${results.summary.fastestFile?.processingTime.toFixed(2)}s)`);
  log(`Slowest:          ${results.summary.slowestFile?.name} (${results.summary.slowestFile?.processingTime.toFixed(2)}s)`);
  log(`\n📁 By File Type:`);
  for (const [ext, stats] of Object.entries(results.summary.byType)) {
    log(`  ${ext}: ${stats.success}/${stats.total} (${(stats.avgTime || 0).toFixed(2)}s avg)`);
  }
  log(`\n💾 Results saved to: ${RESULT_FILE}`);
  log(`📄 Log saved to: ${LOG_FILE}`);
  log('='.repeat(60) + '\n');
}

// Run test
if (require.main === module) {
  runBatchTest().catch(err => {
    log(`\n❌ FATAL ERROR: ${err.message}`);
    log(err.stack);
    process.exit(1);
  });
}

module.exports = { runBatchTest, selectTestFiles };

File: scripts/test_batch_100.py
"""
QSM Batch Test Script - 100 Documents (Python)

Simplified version that directly tests Docling processing
without needing Node.js build artifacts.
"""

import os
import sys
import json
import time
import random
import hashlib
from pathlib import Path
from datetime import datetime
from collections import defaultdict

# Add project root to path
sys.path.insert(0, str(Path(__file__).parent.parent))

try:
    from docling.document_converter import DocumentConverter
    from docling.datamodel.base_models import InputFormat
    DOCLING_AVAILABLE = True
except ImportError:
    print("❌ Error: docling not installed")
    print("Run: pip install docling")
    DOCLING_AVAILABLE = False
    sys.exit(1)

# Configuration
TEST_FOLDER = r"D:\Work\Coding\archi-query-master\Documents"
TARGET_COUNT = 100
RESULT_FILE = Path(__file__).parent.parent / "test_results_100.json"
LOG_FILE = Path(__file__).parent.parent / "test_log_100.txt"

# Results tracking
results = {
    "startTime": datetime.now().isoformat(),
    "testFolder": TEST_FOLDER,
    "targetCount": TARGET_COUNT,
    "totalFiles": 0,
    "processedFiles": 0,
    "failedFiles": 0,
    "totalProcessingTime": 0,
    "avgProcessingTime": 0,
    "files": [],
    "errors": [],
    "summary": {}
}

def log(message):
    """Log to console and file"""
    timestamp = datetime.now().isoformat()
    log_message = f"[{timestamp}] {message}"
    print(message)
    with open(LOG_FILE, 'a', encoding='utf-8') as f:
        f.write(log_message + '\n')

def get_file_hash(filepath):
    """Calculate SHA256 hash of file"""
    hash_sha256 = hashlib.sha256()
    try:
        with open(filepath, 'rb') as f:
            for chunk in iter(lambda: f.read(4096), b""):
                hash_sha256.update(chunk)
        return hash_sha256.hexdigest()
    except Exception as e:
        return None

def select_test_files():
    """Select 100 diverse files using stratified sampling"""
    log(f"📂 Scanning {TEST_FOLDER}...")
    
    # Collect all files by type
    files_by_type = defaultdict(list)
    all_files = []
    
    for root, dirs, files in os.walk(TEST_FOLDER):
        for file in files:
            filepath = os.path.join(root, file)
            ext = os.path.splitext(file)[1].lower()
            
            # Filter supported formats
            if ext in ['.pdf', '.docx', '.doc', '.xlsx', '.xls', '.pptx', '.ppt', '.txt', '.md']:
                try:
                    size = os.path.getsize(filepath)
                    files_by_type[ext].append({
                        'path': filepath,
                        'name': file,
                        'size': size,
                        'ext': ext
                    })
                    all_files.append(files_by_type[ext][-1])
                except Exception as e:
                    log(f"⚠️  Skipping {file}: {e}")
    
    total_available = len(all_files)
    log(f"📊 Found {total_available} documents")
    
    for ext, files in files_by_type.items():
        log(f"  {ext}: {len(files)} files")
    
    # Stratified sampling
    if total_available <= TARGET_COUNT:
        log(f"⚠️  Only {total_available} files available, using all")
        selected = all_files
    else:
        selected = []
        for ext, files in files_by_type.items():
            # Sample proportionally
            proportion = len(files) / total_available
            count_to_sample = max(1, int(TARGET_COUNT * proportion))
            
            # Don't sample more than available
            count_to_sample = min(count_to_sample, len(files))
            
            sampled = random.sample(files, count_to_sample)
            selected.extend(sampled)
            log(f"  Sampled {count_to_sample} {ext} files")
        
        # If we're under target, randomly sample more
        if len(selected) < TARGET_COUNT:
            remaining_files = [f for f in all_files if f not in selected]
            additional = random.sample(remaining_files, min(TARGET_COUNT - len(selected), len(remaining_files)))
            selected.extend(additional)
    
    # Shuffle for random order
    random.shuffle(selected)
    
    log(f"✅ Selected {len(selected)} files for testing\n")
    return selected[:TARGET_COUNT]

def process_file(filepath, index, total):
    """Process single file with Docling"""
    filename = os.path.basename(filepath)
    log(f"[{index}/{total}] Processing: {filename}")
    
    result = {
        'index': index,
        'name': filename,
        'path': filepath,
        'size': os.path.getsize(filepath),
        'ext': os.path.splitext(filename)[1].lower(),
        'hash': get_file_hash(filepath),
        'success': False,
        'processingTime': 0,
        'chunks': 0,
        'wordCount': 0,
        'pageCount': 0,
        'tableCount': 0,
        'error': None
    }
    
    start_time = time.time()
    
    try:
        # Initialize Docling converter with default options
        # Note: backend parameter removed - not supported in current Docling version
        converter = DocumentConverter()  # Auto-detect format and use defaults
        
        # Convert document
        doc_result = converter.convert(filepath)
        
        # Extract metrics
        text = doc_result.document.export_to_markdown()
        result['wordCount'] = len(text.split())
        result['chunks'] = len(text) // 1000  # Rough estimate (1000 chars per chunk)
        
        # Page count (if available)
        if hasattr(doc_result.document, 'pages'):
            result['pageCount'] = len(doc_result.document.pages)
        
        # Table count (if available)
        if hasattr(doc_result.document, 'tables'):
            result['tableCount'] = len(doc_result.document.tables)
        
        result['success'] = True
        log(f"  ✅ Success: {result['wordCount']} words, {result['pageCount']} pages, {result['tableCount']} tables")
        
    except Exception as e:
        result['error'] = str(e)
        log(f"  ❌ Error: {e}")
    
    finally:
        result['processingTime'] = time.time() - start_time
        log(f"  ⏱️  Time: {result['processingTime']:.2f}s\n")
    
    return result

def generate_summary(file_results):
    """Generate test summary statistics"""
    summary = {
        'successRate': 0,
        'avgProcessingTime': 0,
        'fastestFile': None,
        'slowestFile': None,
        'byType': {}
    }
    
    # Overall stats
    successful = [r for r in file_results if r['success']]
    failed = [r for r in file_results if not r['success']]
    
    if file_results:
        summary['successRate'] = f"{len(successful) / len(file_results) * 100:.1f}%"
    
    if successful:
        times = [r['processingTime'] for r in successful]
        summary['avgProcessingTime'] = f"{sum(times) / len(times):.2f}s"
        
        fastest = min(successful, key=lambda x: x['processingTime'])
        slowest = max(successful, key=lambda x: x['processingTime'])
        
        summary['fastestFile'] = {
            'name': fastest['name'],
            'time': f"{fastest['processingTime']:.2f}s"
        }
        summary['slowestFile'] = {
            'name': slowest['name'],
            'time': f"{slowest['processingTime']:.2f}s"
        }
    
    # By file type
    by_type = defaultdict(lambda: {'total': 0, 'success': 0, 'times': []})
    for r in file_results:
        ext = r['ext']
        by_type[ext]['total'] += 1
        if r['success']:
            by_type[ext]['success'] += 1
            by_type[ext]['times'].append(r['processingTime'])
    
    for ext, stats in by_type.items():
        summary['byType'][ext] = {
            'total': stats['total'],
            'success': stats['success'],
            'successRate': f"{stats['success'] / stats['total'] * 100:.1f}%",
            'avgTime': f"{sum(stats['times']) / len(stats['times']):.2f}s" if stats['times'] else "N/A"
        }
    
    return summary

def main():
    """Main test runner"""
    log("🔍 QSM Batch Test - 100 Documents (Python)")
    log("=" * 50)
    
    # Clear log file
    if LOG_FILE.exists():
        LOG_FILE.unlink()
    
    # Select test files
    test_files = select_test_files()
    results['totalFiles'] = len(test_files)
    
    if not test_files:
        log("❌ No files found to test!")
        return
    
    log("🚀 Starting batch processing...\n")
    start_time = time.time()
    
    # Process each file
    for i, file_info in enumerate(test_files, 1):
        result = process_file(file_info['path'], i, len(test_files))
        results['files'].append(result)
        
        if result['success']:
            results['processedFiles'] += 1
        else:
            results['failedFiles'] += 1
            results['errors'].append({
                'file': result['name'],
                'error': result['error']
            })
    
    # Calculate summary
    results['totalProcessingTime'] = time.time() - start_time
    results['avgProcessingTime'] = results['totalProcessingTime'] / len(test_files)
    results['summary'] = generate_summary(results['files'])
    results['endTime'] = datetime.now().isoformat()
    
    # Save results
    with open(RESULT_FILE, 'w', encoding='utf-8') as f:
        json.dump(results, f, indent=2, ensure_ascii=False)
    
    # Print summary
    log("\n" + "=" * 50)
    log("📊 BATCH TEST SUMMARY")
    log("=" * 50)
    log(f"Total Files:     {results['totalFiles']}")
    log(f"Processed:       {results['processedFiles']} ({results['summary']['successRate']})")
    log(f"Failed:          {results['failedFiles']}")
    log(f"Total Time:      {results['totalProcessingTime']:.1f}s ({results['totalProcessingTime']/60:.1f} min)")
    log(f"Avg Time:        {results['summary']['avgProcessingTime']}")
    
    if results['summary'].get('fastestFile'):
        log(f"Fastest:         {results['summary']['fastestFile']['name']} ({results['summary']['fastestFile']['time']})")
    if results['summary'].get('slowestFile'):
        log(f"Slowest:         {results['summary']['slowestFile']['name']} ({results['summary']['slowestFile']['time']})")
    
    log("\nBy Type:")
    for ext, stats in results['summary']['byType'].items():
        log(f"  {ext}: {stats['success']}/{stats['total']} ({stats['successRate']}), avg {stats['avgTime']}")
    
    if results['errors']:
        log(f"\n⚠️  {len(results['errors'])} Errors:")
        for error in results['errors'][:5]:  # Show first 5
            log(f"  - {error['file']}: {error['error'][:100]}")
    
    log(f"\n✅ Results saved to: {RESULT_FILE}")
    log(f"📋 Full log saved to: {LOG_FILE}")
    log("=" * 50)

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        log("\n⚠️  Test interrupted by user")
        sys.exit(1)
    except Exception as e:
        log(f"\n❌ Fatal error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)
